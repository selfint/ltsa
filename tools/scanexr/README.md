# scanexr

Find call path from a known vulnerable value,
to a known controllable value.

A 'value' can be a method, an argument to a method, or anything else.
