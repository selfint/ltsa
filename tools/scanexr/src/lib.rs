use std::{fmt::Debug, path::Path};

use anyhow::{Context, Result};
use async_recursion::async_recursion;
use async_trait::async_trait;
use lsp_client::client::Client;
use step::Step;
use tree_sitter::{Language, Query};

use crate::utils::get_query_steps;

pub mod step;
pub mod tracers;
pub mod utils;

pub type Stacktrace<C> = Vec<Step<C>>;

#[async_trait]
pub trait Tracer: Send + Sync {
    type StepContext: Debug + Default + Clone + Send + Sync;

    fn get_language(&self) -> Language;

    /// given a step, get all possible stack traces leading to it
    async fn get_stacktraces(
        &self,
        lsp_client: &Client,
        step: &Step<Self::StepContext>,
        stop_at: &[Step<Self::StepContext>],
    ) -> Result<Vec<Stacktrace<Self::StepContext>>>;
}

pub async fn get_all_stacktraces<T: Tracer>(
    tracer: &T,
    lsp_client: &Client,
    root_dir: &Path,
    pub_queries: &[(Query, u32)],
    hacky_query: &(Query, u32),
) -> Result<Vec<Stacktrace<T::StepContext>>> {
    let mut pub_steps = vec![];
    for pub_query in pub_queries {
        let steps = get_query_steps(root_dir, tracer.get_language(), pub_query)
            .context("getting pub query steps")?;
        for step in steps {
            if !pub_steps.contains(&step) {
                pub_steps.push(step);
            }
        }
    }

    let hacky_steps = get_query_steps(root_dir, tracer.get_language(), hacky_query)
        .context("getting hacky steps")?;

    let mut all_stacktraces = vec![];
    for hacky_step in &hacky_steps {
        let stacktraces = _get_all_stacktraces(tracer, lsp_client, hacky_step, &pub_steps)
            .await
            .context("completing stacktraces")?;

        all_stacktraces.extend(stacktraces);
    }

    Ok(all_stacktraces)
}

#[async_recursion]
async fn _get_all_stacktraces<T: Tracer>(
    tracer: &T,
    lsp_client: &Client,
    step: &Step<T::StepContext>,
    stop_at: &[Step<T::StepContext>],
) -> Result<Vec<Stacktrace<T::StepContext>>> {
    // get stacktraces leading to step
    let stacktraces = tracer.get_stacktraces(lsp_client, step, stop_at).await?;

    // complete stacktraces leading to step
    let mut completed_stacktraces = vec![];
    for stacktrace in stacktraces {
        let next_stacktraces = _get_all_stacktraces(tracer, lsp_client, step, stop_at).await?;
        for next_stacktrace in next_stacktraces {
            let mut completed_stacktrace = stacktrace.clone();
            completed_stacktrace.extend(next_stacktrace);

            completed_stacktraces.push(completed_stacktrace);
        }
    }

    Ok(completed_stacktraces)
}
