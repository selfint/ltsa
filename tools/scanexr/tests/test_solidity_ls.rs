use scanexr::{
    language_provider::LspProvider, languages::solidity::SolidityLs, test_utils::display_locations,
    utils::visit_dirs,
};

#[tokio::test]
async fn test_find_definitions() {
    macro_rules! test_definitions {
        ($input:literal) => {
            let (root_dir, location, _, _) = scanexr::test_utils::setup_test_dir($input);

            let mut project_files = vec![];
            visit_dirs(root_dir.path(), &mut |f| project_files.push(f.path()))
                .expect("failed to get project files");

            let lsp = SolidityLs::new(root_dir.path(), project_files)
                .await
                .expect("failed to start solidity ls");

            let definitions = lsp
                .find_definitions(&location)
                .await
                .expect("failed to find definition");

            let definitions = definitions.into_iter().map(|d| (d, ())).collect();
            let snapshot = format!(
                r#"
### input ###
{}
### output ###
{}"#,
                $input,
                display_locations::<()>(definitions),
            );

            insta::assert_snapshot!(snapshot);
        };
    }

    test_definitions!(
        r#"
contract.sol
#@#
pragma solidity ^0.8.19;

contract Contract {
    function withdraw() public {
        address target = msg.sender;

        (bool sent, ) = target.call{value: 1}("");
                     // ^^^^^^ start
    }
}
        "#
    );

    test_definitions!(
        r#"
contract.sol
#@#
pragma solidity ^0.8.19;

contract Contract {
    function foo() public {}

    function withdraw() public {
        address target = msg.sender;
        
        foo();
    //  ^^^ start

        (bool sent, ) = target.call{value: 1}("");
    }
}
        "#
    );
}

#[tokio::test]
async fn test_find_references() {
    macro_rules! test_references {
        ($input:literal) => {
            let (root_dir, location, _, _) = scanexr::test_utils::setup_test_dir($input);

            let mut project_files = vec![];
            visit_dirs(root_dir.path(), &mut |f| project_files.push(f.path()))
                .expect("failed to get project files");

            let lsp = SolidityLs::new(root_dir.path(), project_files)
                .await
                .expect("failed to start solidity ls");

            let references = lsp
                .find_references(&location)
                .await
                .expect("failed to find references");

            let references = references.into_iter().map(|d| (d, ())).collect();
            let snapshot = format!(
                r#"
### input ###
{}
### output ###
{}"#,
                $input,
                display_locations::<()>(references),
            );

            insta::assert_snapshot!(snapshot);
        };
    }

    test_references!(
        r#"
contract.sol
#@#
pragma solidity ^0.8.19;

contract Contract {
    function foo() public {}
         //  ^^^ start

    function withdraw() public {
        address target = msg.sender;
        
        foo();
        foo();

        (bool sent, ) = target.call{value: 1}("");
    }
}
        "#
    );
}
