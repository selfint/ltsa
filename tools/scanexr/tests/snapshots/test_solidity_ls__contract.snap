---
source: tools/scanexr/tests/test_solidity_ls.rs
expression: snapshot
---
input:

contract.sol
#@#
// SPDX-License-Identifier: MIT
pragma solidity ^0.8.17;

import {other, hacky} from "./other_file.sol";

/*
EtherStore is a contract where you can deposit and withdraw ETH.
This contract is vulnerable to re-entrancy attack.
Let's see why.

1. Deploy EtherStore
2. Deposit 1 Ether each from Account 1 (Alice) and Account 2 (Bob) into EtherStore
3. Deploy Attack with address of EtherStore
4. Call Attack.attack sending 1 ether (using Account 3 (Eve)).
   You will get 3 Ethers back (2 Ether stolen from Alice and Bob,
   plus 1 Ether sent from this contract).

What happened?
Attack was able to call EtherStore.withdraw multiple times before
EtherStore.withdraw finished executing.

Here is how the functions were called
- Attack.attack
- EtherStore.deposit
- EtherStore.withdraw
- Attack fallback (receives 1 Ether)
- EtherStore.withdraw
- Attack.fallback (receives 1 Ether)
- EtherStore.withdraw
- Attack fallback (receives 1 Ether)
*/

contract EtherStore {
    mapping(address => uint) public balances;

    function deposit() public payable {
        balances[msg.sender] += msg.value;
    }

    function getSender2() private view returns (address) {
        return msg.sender;
        /*     ^^^^^^^^^^ definition */
    }

    function getSender() private view returns (address) {
        if (false) {
            return foo(0x0000000000000000000000000000000000000000, msg.sender);
            /*                                                     ^^^^^^^^^^ definition */
        } else if (false) {
            return ret2(0x0000000000000000000000000000000000000000, msg.sender);
            /*                                                      ^^^^^^^^^^ definition */
        }

        return getSender2();
    }

    function ret2(address a, address a2) private pure returns (address) {
        (uint c, address d) = other(a2, 1);

        return d;
    }

    function foo(address a, address b) private pure returns (address) {
        return a;
    }

    function withdraw() public {
        address sender = getSender();

        address bar = foo(sender, sender);

        uint bal = balances[bar];
        require(bal > 0);

        hacky(bar, bal);

        balances[bar] = 0;
    }

    // Helper function to check the balance of this contract
    function getBalance() public view returns (uint) {
        return address(this).balance;
    }
}

---
other_file.sol
#@#
// SPDX-License-Identifier: MIT
pragma solidity ^0.8.17;

function other(address a, uint b) pure returns (uint, address) {
    return (b, a);
}

function hacky(address target, uint amount) {
    (bool sent, ) = target.call{value: amount}("");
    /*              ^^^^^^ start */
    require(sent, "Failed to send Ether");
}

    
output:Path: 0
other_file.sol
#@#
}

function hacky(address target, uint amount) {
    (bool sent, ) = target.call{value: amount}("");
                    ^^^^^^ Meta: ()
    /*              ^^^^^^ start */
    require(sent, "Failed to send Ether");
}
---
other_file.sol
#@#
}

function hacky(address target, uint amount) {
    (bool sent, ) = target.call{value: amount}("");
                    ^^^^^^ Meta: ()
    /*              ^^^^^^ start */
    require(sent, "Failed to send Ether");
}
---
other_file.sol
#@#
}

function hacky(address target, uint amount) {
    (bool sent, ) = target.call{value: amount}("");
                    ^^^^^^ Meta: ()
    /*              ^^^^^^ start */
    require(sent, "Failed to send Ether");
}
---
other_file.sol
#@#
    return (b, a);
}

function hacky(address target, uint amount) {
                       ^^^^^^ Meta: ()
    (bool sent, ) = target.call{value: amount}("");
    /*              ^^^^^^ start */
    require(sent, "Failed to send Ether");
---
other_file.sol
#@#
    return (b, a);
}

function hacky(address target, uint amount) {
         ^^^^^ Meta: ()
    (bool sent, ) = target.call{value: amount}("");
    /*              ^^^^^^ start */
    require(sent, "Failed to send Ether");
---
contract.sol
#@#
        uint bal = balances[bar];
        require(bal > 0);

        hacky(bar, bal);
        ^^^^^ Meta: ()

        balances[bar] = 0;
    }
---
contract.sol
#@#
        uint bal = balances[bar];
        require(bal > 0);

        hacky(bar, bal);
              ^^^ Meta: ()

        balances[bar] = 0;
    }
---
contract.sol
#@#
        uint bal = balances[bar];
        require(bal > 0);

        hacky(bar, bal);
              ^^^ Meta: ()

        balances[bar] = 0;
    }
---
contract.sol
#@#
    function withdraw() public {
        address sender = getSender();

        address bar = foo(sender, sender);
                ^^^ Meta: ()

        uint bal = balances[bar];
        require(bal > 0);
---
contract.sol
#@#
    function withdraw() public {
        address sender = getSender();

        address bar = foo(sender, sender);
        ^^^^^^^^^^^ Meta: ()

        uint bal = balances[bar];
        require(bal > 0);
---
contract.sol
#@#
    function withdraw() public {
        address sender = getSender();

        address bar = foo(sender, sender);
                      ^^^^^^^^^^^^^^^^^^^ Meta: ()

        uint bal = balances[bar];
        require(bal > 0);
---
contract.sol
#@#
    function withdraw() public {
        address sender = getSender();

        address bar = foo(sender, sender);
                      ^^^ Meta: ()

        uint bal = balances[bar];
        require(bal > 0);
---
contract.sol
#@#
        return d;
    }

    function foo(address a, address b) private pure returns (address) {
             ^^^ Meta: ()
        return a;
    }

---
contract.sol
#@#
    }

    function foo(address a, address b) private pure returns (address) {
        return a;
               ^ Meta: ()
    }

    function withdraw() public {
---
contract.sol
#@#
    }

    function foo(address a, address b) private pure returns (address) {
        return a;
               ^ Meta: ()
    }

    function withdraw() public {
---
contract.sol
#@#
        return d;
    }

    function foo(address a, address b) private pure returns (address) {
                         ^ Meta: ()
        return a;
    }

---
contract.sol
#@#
    function withdraw() public {
        address sender = getSender();

        address bar = foo(sender, sender);
                      ^^^ Meta: ()

        uint bal = balances[bar];
        require(bal > 0);
---
contract.sol
#@#
    function withdraw() public {
        address sender = getSender();

        address bar = foo(sender, sender);
                          ^^^^^^ Meta: ()

        uint bal = balances[bar];
        require(bal > 0);
---
contract.sol
#@#
    function withdraw() public {
        address sender = getSender();

        address bar = foo(sender, sender);
                          ^^^^^^ Meta: ()

        uint bal = balances[bar];
        require(bal > 0);
---
contract.sol
#@#
    }

    function withdraw() public {
        address sender = getSender();
                ^^^^^^ Meta: ()

        address bar = foo(sender, sender);

---
contract.sol
#@#
    }

    function withdraw() public {
        address sender = getSender();
        ^^^^^^^^^^^^^^ Meta: ()

        address bar = foo(sender, sender);

---
contract.sol
#@#
    }

    function withdraw() public {
        address sender = getSender();
                         ^^^^^^^^^^^ Meta: ()

        address bar = foo(sender, sender);

---
contract.sol
#@#
    }

    function withdraw() public {
        address sender = getSender();
                         ^^^^^^^^^ Meta: ()

        address bar = foo(sender, sender);

---
contract.sol
#@#
        /*     ^^^^^^^^^^ definition */
    }

    function getSender() private view returns (address) {
             ^^^^^^^^^ Meta: ()
        if (false) {
            return foo(0x0000000000000000000000000000000000000000, msg.sender);
            /*                                                     ^^^^^^^^^^ definition */
---
contract.sol
#@#
            return foo(0x0000000000000000000000000000000000000000, msg.sender);
            /*                                                     ^^^^^^^^^^ definition */
        } else if (false) {
            return ret2(0x0000000000000000000000000000000000000000, msg.sender);
                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ Meta: ()
            /*                                                      ^^^^^^^^^^ definition */
        }

---
contract.sol
#@#
            return foo(0x0000000000000000000000000000000000000000, msg.sender);
            /*                                                     ^^^^^^^^^^ definition */
        } else if (false) {
            return ret2(0x0000000000000000000000000000000000000000, msg.sender);
                   ^^^^ Meta: ()
            /*                                                      ^^^^^^^^^^ definition */
        }

---
contract.sol
#@#
        return getSender2();
    }

    function ret2(address a, address a2) private pure returns (address) {
             ^^^^ Meta: ()
        (uint c, address d) = other(a2, 1);

        return d;
---
contract.sol
#@#
    function ret2(address a, address a2) private pure returns (address) {
        (uint c, address d) = other(a2, 1);

        return d;
               ^ Meta: ()
    }

    function foo(address a, address b) private pure returns (address) {
---
contract.sol
#@#
    function ret2(address a, address a2) private pure returns (address) {
        (uint c, address d) = other(a2, 1);

        return d;
               ^ Meta: ()
    }

    function foo(address a, address b) private pure returns (address) {
---
contract.sol
#@#
    }

    function ret2(address a, address a2) private pure returns (address) {
        (uint c, address d) = other(a2, 1);
                         ^ Meta: ()

        return d;
    }
---
contract.sol
#@#
    }

    function ret2(address a, address a2) private pure returns (address) {
        (uint c, address d) = other(a2, 1);
        ^^^^^^^^^^^^^^^^^^^ Meta: ()

        return d;
    }
---
contract.sol
#@#
    }

    function ret2(address a, address a2) private pure returns (address) {
        (uint c, address d) = other(a2, 1);
                              ^^^^^^^^^^^^ Meta: ()

        return d;
    }
---
contract.sol
#@#
    }

    function ret2(address a, address a2) private pure returns (address) {
        (uint c, address d) = other(a2, 1);
                              ^^^^^ Meta: ()

        return d;
    }
---
other_file.sol
#@#
// SPDX-License-Identifier: MIT
pragma solidity ^0.8.17;

function other(address a, uint b) pure returns (uint, address) {
         ^^^^^ Meta: ()
    return (b, a);
}

---
other_file.sol
#@#
pragma solidity ^0.8.17;

function other(address a, uint b) pure returns (uint, address) {
    return (b, a);
           ^^^^^^ Meta: ()
}

function hacky(address target, uint amount) {
---
other_file.sol
#@#
pragma solidity ^0.8.17;

function other(address a, uint b) pure returns (uint, address) {
    return (b, a);
               ^ Meta: ()
}

function hacky(address target, uint amount) {
---
other_file.sol
#@#
pragma solidity ^0.8.17;

function other(address a, uint b) pure returns (uint, address) {
    return (b, a);
               ^ Meta: ()
}

function hacky(address target, uint amount) {
---
other_file.sol
#@#
// SPDX-License-Identifier: MIT
pragma solidity ^0.8.17;

function other(address a, uint b) pure returns (uint, address) {
                       ^ Meta: ()
    return (b, a);
}

---
contract.sol
#@#
    }

    function ret2(address a, address a2) private pure returns (address) {
        (uint c, address d) = other(a2, 1);
                              ^^^^^ Meta: ()

        return d;
    }
---
contract.sol
#@#
    }

    function ret2(address a, address a2) private pure returns (address) {
        (uint c, address d) = other(a2, 1);
                                    ^^ Meta: ()

        return d;
    }
---
contract.sol
#@#
    }

    function ret2(address a, address a2) private pure returns (address) {
        (uint c, address d) = other(a2, 1);
                                    ^^ Meta: ()

        return d;
    }
---
contract.sol
#@#
        return getSender2();
    }

    function ret2(address a, address a2) private pure returns (address) {
                                     ^^ Meta: ()
        (uint c, address d) = other(a2, 1);

        return d;
---
contract.sol
#@#
        return getSender2();
    }

    function ret2(address a, address a2) private pure returns (address) {
             ^^^^ Meta: ()
        (uint c, address d) = other(a2, 1);

        return d;
---
contract.sol
#@#
            return foo(0x0000000000000000000000000000000000000000, msg.sender);
            /*                                                     ^^^^^^^^^^ definition */
        } else if (false) {
            return ret2(0x0000000000000000000000000000000000000000, msg.sender);
                   ^^^^ Meta: ()
            /*                                                      ^^^^^^^^^^ definition */
        }

---
contract.sol
#@#
            return foo(0x0000000000000000000000000000000000000000, msg.sender);
            /*                                                     ^^^^^^^^^^ definition */
        } else if (false) {
            return ret2(0x0000000000000000000000000000000000000000, msg.sender);
                                                                    ^^^^^^^^^^ Meta: ()
            /*                                                      ^^^^^^^^^^ definition */
        }

---
Path: 1
other_file.sol
#@#
}

function hacky(address target, uint amount) {
    (bool sent, ) = target.call{value: amount}("");
                    ^^^^^^ Meta: ()
    /*              ^^^^^^ start */
    require(sent, "Failed to send Ether");
}
---
other_file.sol
#@#
}

function hacky(address target, uint amount) {
    (bool sent, ) = target.call{value: amount}("");
                    ^^^^^^ Meta: ()
    /*              ^^^^^^ start */
    require(sent, "Failed to send Ether");
}
---
other_file.sol
#@#
}

function hacky(address target, uint amount) {
    (bool sent, ) = target.call{value: amount}("");
                    ^^^^^^ Meta: ()
    /*              ^^^^^^ start */
    require(sent, "Failed to send Ether");
}
---
other_file.sol
#@#
    return (b, a);
}

function hacky(address target, uint amount) {
                       ^^^^^^ Meta: ()
    (bool sent, ) = target.call{value: amount}("");
    /*              ^^^^^^ start */
    require(sent, "Failed to send Ether");
---
other_file.sol
#@#
    return (b, a);
}

function hacky(address target, uint amount) {
         ^^^^^ Meta: ()
    (bool sent, ) = target.call{value: amount}("");
    /*              ^^^^^^ start */
    require(sent, "Failed to send Ether");
---
contract.sol
#@#
        uint bal = balances[bar];
        require(bal > 0);

        hacky(bar, bal);
        ^^^^^ Meta: ()

        balances[bar] = 0;
    }
---
contract.sol
#@#
        uint bal = balances[bar];
        require(bal > 0);

        hacky(bar, bal);
              ^^^ Meta: ()

        balances[bar] = 0;
    }
---
contract.sol
#@#
        uint bal = balances[bar];
        require(bal > 0);

        hacky(bar, bal);
              ^^^ Meta: ()

        balances[bar] = 0;
    }
---
contract.sol
#@#
    function withdraw() public {
        address sender = getSender();

        address bar = foo(sender, sender);
                ^^^ Meta: ()

        uint bal = balances[bar];
        require(bal > 0);
---
contract.sol
#@#
    function withdraw() public {
        address sender = getSender();

        address bar = foo(sender, sender);
        ^^^^^^^^^^^ Meta: ()

        uint bal = balances[bar];
        require(bal > 0);
---
contract.sol
#@#
    function withdraw() public {
        address sender = getSender();

        address bar = foo(sender, sender);
                      ^^^^^^^^^^^^^^^^^^^ Meta: ()

        uint bal = balances[bar];
        require(bal > 0);
---
contract.sol
#@#
    function withdraw() public {
        address sender = getSender();

        address bar = foo(sender, sender);
                      ^^^ Meta: ()

        uint bal = balances[bar];
        require(bal > 0);
---
contract.sol
#@#
        return d;
    }

    function foo(address a, address b) private pure returns (address) {
             ^^^ Meta: ()
        return a;
    }

---
contract.sol
#@#
    }

    function foo(address a, address b) private pure returns (address) {
        return a;
               ^ Meta: ()
    }

    function withdraw() public {
---
contract.sol
#@#
    }

    function foo(address a, address b) private pure returns (address) {
        return a;
               ^ Meta: ()
    }

    function withdraw() public {
---
contract.sol
#@#
        return d;
    }

    function foo(address a, address b) private pure returns (address) {
                         ^ Meta: ()
        return a;
    }

---
contract.sol
#@#
    function withdraw() public {
        address sender = getSender();

        address bar = foo(sender, sender);
                      ^^^ Meta: ()

        uint bal = balances[bar];
        require(bal > 0);
---
contract.sol
#@#
    function withdraw() public {
        address sender = getSender();

        address bar = foo(sender, sender);
                          ^^^^^^ Meta: ()

        uint bal = balances[bar];
        require(bal > 0);
---
contract.sol
#@#
    function withdraw() public {
        address sender = getSender();

        address bar = foo(sender, sender);
                          ^^^^^^ Meta: ()

        uint bal = balances[bar];
        require(bal > 0);
---
contract.sol
#@#
    }

    function withdraw() public {
        address sender = getSender();
                ^^^^^^ Meta: ()

        address bar = foo(sender, sender);

---
contract.sol
#@#
    }

    function withdraw() public {
        address sender = getSender();
        ^^^^^^^^^^^^^^ Meta: ()

        address bar = foo(sender, sender);

---
contract.sol
#@#
    }

    function withdraw() public {
        address sender = getSender();
                         ^^^^^^^^^^^ Meta: ()

        address bar = foo(sender, sender);

---
contract.sol
#@#
    }

    function withdraw() public {
        address sender = getSender();
                         ^^^^^^^^^ Meta: ()

        address bar = foo(sender, sender);

---
contract.sol
#@#
        /*     ^^^^^^^^^^ definition */
    }

    function getSender() private view returns (address) {
             ^^^^^^^^^ Meta: ()
        if (false) {
            return foo(0x0000000000000000000000000000000000000000, msg.sender);
            /*                                                     ^^^^^^^^^^ definition */
---
contract.sol
#@#
            /*                                                      ^^^^^^^^^^ definition */
        }

        return getSender2();
               ^^^^^^^^^^^^ Meta: ()
    }

    function ret2(address a, address a2) private pure returns (address) {
---
contract.sol
#@#
            /*                                                      ^^^^^^^^^^ definition */
        }

        return getSender2();
               ^^^^^^^^^^ Meta: ()
    }

    function ret2(address a, address a2) private pure returns (address) {
---
contract.sol
#@#
        balances[msg.sender] += msg.value;
    }

    function getSender2() private view returns (address) {
             ^^^^^^^^^^ Meta: ()
        return msg.sender;
        /*     ^^^^^^^^^^ definition */
    }
---
contract.sol
#@#
    }

    function getSender2() private view returns (address) {
        return msg.sender;
               ^^^^^^^^^^ Meta: ()
        /*     ^^^^^^^^^^ definition */
    }

